src/status_im2/contexts/chat/lightbox/zoomable_image/view.cljs

```diff
@@ -0,0 +1,313 @@
+  ([value duration]
+   (reanimated/with-timing-duration value duration)))
+
+(defn set-val-decay
+  [animation value bounds]
+  (reanimated/animate-shared-value-with-decay animation value bounds))
+
+;; MATH
+(defn get-max-offset
+  [size screen-size scale]
+  (/ (/ (- (* size scale) screen-size) 2) scale))
+
+(defn get-scale-ratio
+  [new-scale saved-scale]
+  (/ (- (dec new-scale) (dec saved-scale)) new-scale))
+
+(defn get-current-center
+  [size scaled-size offset]
+  (- (+ (/ (- size scaled-size) 2) (/ scaled-size 2)) offset))
+
+;; 5 Gestures: tap, double-tap, pinch, pan-x, pan-y
+(defn tap-gesture
+  [on-tap]
+  (->
+    (gesture/gesture-tap)
+    (gesture/on-start #(on-tap))))
+
+(defn double-tap-gesture
+  [{:keys [width height screen-height]}
+   {:keys [scale pan-x pan-x-start pan-y pan-y-start]}
+   {:keys [y-threshold-scale]}
+   rescale]
+  (->
+    (gesture/gesture-tap)
+    (gesture/number-of-taps 2)
+    (gesture/on-start (fn [evt]
+                        ;; Scale to x2
+                        (if (= (get-val scale) min-scale)
+                          ;; Find translate-x value
+                          (let [focal-x     (- (/ width 2) (.-x evt))
+                                max-pan-x   (get-max-offset width width double-tap-scale)
+                                translate-x (min (Math/abs focal-x) max-pan-x)
+                                translate-x (if (neg? focal-x) (- translate-x) translate-x)
+                                ;; Find translate-y value
+                                focal-y     (- (/ height 2) (.-y evt))
+                                max-pan-y   (get-max-offset height screen-height double-tap-scale)
+                                translate-y (min (Math/abs focal-y) max-pan-y)
+                                translate-y (if (neg? focal-y) (- translate-y) translate-y)]
+                            ;; apply animations
+                            (set-val pan-x (timing translate-x))
+                            (set-val pan-x-start (timing translate-x))
+                            ;; animate y position only if the double-tap-scale exceeds the threshold
+                            (when (> double-tap-scale y-threshold-scale)
+                              (set-val pan-y (timing translate-y))
+                              (set-val pan-y-start (timing translate-y)))
+                            (rescale double-tap-scale))
+                          ;; Otherwise scale back to x1
+                          (rescale min-scale))))))
+
+(defn pinch-gesture
+  [{:keys [width height]}
+   {:keys [saved-scale scale pinch-x pinch-y pinch-x-start pinch-y-start pan-x-start pan-y-start]}
+   {:keys [pan-x-enabled? pan-y-enabled? x-threshold-scale y-threshold-scale focal-x focal-y]}
+   rescale]
+  (->
+    (gesture/gesture-pinch)
+    (gesture/on-start (fn [evt]
+                        ;; Read the x and y touch positions on the screen
+                        (reset! focal-x (.-focalX evt))
```

Comment:
The lack of type hints will generate warnings and there's always a chance that the release build can have subtle bugs because of this. The advanced compilation mode can get confused. I've faced this issue before in status-mobile and it's a nightmare to trace back the root cause if you don't know it's because of a lack of type hint. So we better always add the `^js` type hints in the correct places when doing interop.

But forget what I just said, it's even better if you can stick to `oops/oget`, it's in the guidelines ;)

Reply:
Marking with `^js` seems to be more `concise`. Is there an advantage to using `oops/oget` over `^js`?

Reply:
@ilmotta Do we mark the top `evt` in the function argument, or also every single `evt`?

Reply:
Good questions :)

>Do we mark the top evt in the function argument, or also every single evt?

We need to type hint only once per function body. If a function body has interop macros, like `(.-focalX evt)`, then the function argument should add the type hint `^js evt` to make sure the advanced compilation to Javascript does its best.

But we'll see plenty of cases where the type hint should be next to the Javascript object, for example when we need to deref and call a native Javascript method. In such cases, the type hint might need to be added multiple times.

```clojure
(.scrollToIndex ^js @scroll-ref)
```

>Marking with ^js seems to be more concise. Is there an advantage to using oops/oget over ^js?

I agree, interop macros can be more concise, but the team has decided on using `oops`. oops offers some advantages, like:

- Less risk of null/undefined exceptions when doing chaining properties. `oops/oget` gives this for free, sort of similar to `m.z?.a`, a new syntax already supported in major browsers.
- When using oops, the type hint is unnecessary. I think the macros will add them for us.
- oops also helps generate valid code when advanced optimizations are enabled (which they are in status-mobile). In some cases the compiler might rename some references and the code can break.

About conciseness, sometimes it might be good to have some exceptions about namespace aliasing. If oops ends up being so prevalent in the codebase, we could consider only using `:refer [oget]`, that should make things less fatiguing.

Reply:
Thanks @ilmotta for the information. Refactored it to using `oget`